flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/TypeInfoDataTypeConverter.java

```diff
@@ -260,35 +261,70 @@ private static DataType convertToStructuredType(
                                         fieldNames[pos],
                                         toDataType(dataTypeFactory, compositeType.getTypeAt(pos))));
 
-        final List<String> fieldNamesOrdered;
+        final List<String> fieldNamesReordered;
         final boolean isNullable;
         // for POJOs and Avro records
         if (compositeType instanceof PojoTypeInfo) {
-            // POJO serializer supports top-level nulls
-            isNullable = true;
-            // all fields are nullable
-            fieldDataTypes.replaceAll((name, dataType) -> dataType.nullable());
-
             final PojoTypeInfo<?> pojoTypeInfo = (PojoTypeInfo<?>) compositeType;
-            final List<Field> fields =
+            final List<Field> pojoFields =
                     IntStream.range(0, arity)
                             .mapToObj(pojoTypeInfo::getPojoFieldAt)
                             .map(PojoField::getField)
                             .collect(Collectors.toList());
+
+            // POJO serializer supports top-level nulls
+            isNullable = true;
+
+            // based on type information all fields are boxed classes,
+            // therefore we need to check the reflective field for more details
+            fieldDataTypes.replaceAll(
+                    (name, dataType) -> {
+                        final Class<?> fieldClass =
+                                pojoFields.stream()
+                                        .filter(f -> f.getName().equals(name))
+                                        .findFirst()
+                                        .orElseThrow(IllegalArgumentException::new)
```

Comment:
nit: I'd rather use `IllegalStateException` as it is more likely an internal Flink error that the `TypeInformation` does not reflect its class.